src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp

```diff
@@ -7792,6 +7792,231 @@ struct TransposeReduceSimplify : public OpRewritePattern<stablehlo::ReduceOp> {
+      }
+    }
+
+    return success();
+  }
+};
+
+// (mul (sign x) (abs x)) -> x
+// (mul (abs x) (sign x)) -> x
+struct SignAbsSimplify : public OpRewritePattern<stablehlo::MulOp> {
+  using OpRewritePattern<stablehlo::MulOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(stablehlo::MulOp op,
+                                PatternRewriter &rewriter) const override {
+    auto lhs = op.getOperand(0);
+    auto rhs = op.getOperand(1);
+
+    auto lhsSignOp = lhs.getDefiningOp<stablehlo::SignOp>();
+    if (lhsSignOp) {
+      auto rhsAbsOp = rhs.getDefiningOp<stablehlo::AbsOp>();
+      if (!rhsAbsOp)
+        return failure();
+
+      if (lhsSignOp.getOperand() != rhsAbsOp.getOperand())
+        return failure();
+
+      rewriter.replaceOp(op, lhsSignOp.getOperand());
+      return success();
+    }
+
+    auto rhsSignOp = rhs.getDefiningOp<stablehlo::SignOp>();
+    if (rhsSignOp) {
+      auto lhsAbsOp = lhs.getDefiningOp<stablehlo::AbsOp>();
+      if (!lhsAbsOp)
+        return failure();
+
+      if (rhsSignOp.getOperand() != lhsAbsOp.getOperand())
+        return failure();
+
+      rewriter.replaceOp(op, rhsSignOp.getOperand());
+      return success();
+    }
+
+    return failure();
+  }
+};
+
+// (mul (neg x) (neg y)) -> (mul x y)
+// (mul (neg x) y) -> (neg (mul x y))
+// (mul x (neg y)) -> (neg (mul x y))
+struct MultiplyNegateSimplify : public OpRewritePattern<stablehlo::MulOp> {
+  using OpRewritePattern<stablehlo::MulOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(stablehlo::MulOp op,
+                                PatternRewriter &rewriter) const override {
+    auto lhs = op.getOperand(0);
+    auto rhs = op.getOperand(1);
+
+    auto lhsNegOp = lhs.getDefiningOp<stablehlo::NegOp>();
+    auto rhsNegOp = rhs.getDefiningOp<stablehlo::NegOp>();
+    if (!lhsNegOp && !rhsNegOp)
+      return failure();
+
+    if (lhsNegOp) {
+      if (rhsNegOp) {
+        rewriter.replaceOpWithNewOp<stablehlo::MulOp>(op, lhsNegOp.getOperand(),
+                                                      rhsNegOp.getOperand());
+        return success();
+      } else {
+        if (!isOnlyUsedInOperation(lhsNegOp, op))
+          return failure();
+        auto newOp = rewriter.create<stablehlo::MulOp>(
+            op.getLoc(), lhsNegOp.getOperand(), rhs);
+        rewriter.replaceOpWithNewOp<stablehlo::NegOp>(op, newOp);
+        return success();
+      }
+    } else if (rhsNegOp) {
+      if (!isOnlyUsedInOperation(rhsNegOp, op))
+        return failure();
+      auto newOp = rewriter.create<stablehlo::MulOp>(op.getLoc(), lhs,
+                                                     rhsNegOp.getOperand());
+      rewriter.replaceOpWithNewOp<stablehlo::NegOp>(op, newOp);
+      return success();
+    }
+
+    return failure();
+  }
+};
+
+// This pattern only does partially the following. We rely on transforming the op to a
+// pattern which further uses the above pattern.
+// (mul (sign x) (add (abs x) (abs x))) -> (mul x x)
```

Comment:
longer term i feel like this merits a broader sign analysis (alongside perhaps a transpose analysis)